Handle candidates which arrive before remote description

With the AWS IoT signaling backend, it looks like candidates are occasionally delivered before the remote description, even though the remote description is sent before the first candidate according to the debug messages in the Javascript console:

Javascript console:
```
LL: onCreateOfferSuccess
index.es.js:8536 LL: Sending to device endpoint: offer
index.es.js:8626 LL: New signaling state: have-local-offer
index.es.js:8622 LL: New ICE gathering state: gathering
index.es.js:8628 LL: New local ICE candidate
index.es.js:8628 LL: local ICE candidate: {  "candidate": "candidate:4079541279 1 udp 2113937151 192.168.1.74 53084 typ host generation 0 ufrag 8VJI network-cost 999",  "sdpMid": "video",  "sdpMLineIndex": 0,  "usernameFragment": "8VJI"}
index.es.js:8536 LL: Sending to device endpoint: iceCandidate
```

RPi console:
```
2019/01/25 19:57:48.738143 client.go:126: Received message on topic: dev-sC/a-b88adf47-e018-4b88-a174-278bfa9bb22a-qaEjy$prod/$/apple/a-b88adf47-e018-4b88-a174-278bfa9bb22a-qaEjy=6nixsvgasbkcdjnd0mz/1548474342-@/
2019/01/25 19:57:48.752399 main.go:109: Remote ICE candidate:4079541279 1 udp 2113937151 192.168.1.74 53084 typ host generation 0 ufrag 8VJI network-cost 999
panic: Must call SetRemoteDescription before AddRemoteCandidate
```
Seems like the fix here would be to buffer candidates (with some expiration time) on the device side?

--- a/internal/ice/agent.go
+++ b/internal/ice/agent.go
@@ -1,6 +1,7 @@
 package ice
 
 import (
+	"errors"
 	"fmt"
 	"log"
 	"net"
@@ -29,17 +30,24 @@ type Agent struct {
 }
 
 // Create a new ICE agent with the given username and passwords.
-func NewAgent(username, localPassword, remotePassword string) *Agent {
-	a := &Agent{}
+func NewAgent() *Agent {
+	return &Agent{
+		ready: make(chan *ChannelConn, 1),
+	}
+}
+
+func (a *Agent) Configure(username, localPassword, remotePassword string) {
 	a.username = username
 	a.localPassword = localPassword
 	a.remotePassword = remotePassword
-	a.ready = make(chan *ChannelConn, 1)
-	return a
 }
 
 // On success, returns a net.Conn object from which data can be read/written.
 func (a *Agent) EstablishConnection(lcand chan<- string) (net.Conn, error) {
+	if a.username == "" {
+		return nil, errors.New("ICE agent not configured")
+	}
+
 	// TODO: Support multiple components
 	//components := []int{1}
 	component := 1

--- a/peer_connection.go
+++ b/peer_connection.go
@@ -57,6 +57,7 @@ func NewPeerConnection() *PeerConnection {
 	}
 
 	pc := &PeerConnection{
+		iceAgent:     ice.NewAgent(),
 		certPEMBlock: cert,
 		keyPEMBlock:  key,
 		fingerprint:  fp,
@@ -139,10 +140,6 @@ func (pc *PeerConnection) createAnswer() sdp.Session {
 
 // Set remote SDP offer. Return SDP answer.
 func (pc *PeerConnection) SetRemoteDescription(sdpOffer string) (sdpAnswer string, err error) {
-	if pc.iceAgent != nil {
-		panic("SetRemoteDescription must only be called once")
-	}
-
 	offer, err := sdp.ParseSession(sdpOffer)
 	if err != nil {
 		return
@@ -156,7 +153,7 @@ func (pc *PeerConnection) SetRemoteDescription(sdpOffer string) (sdpAnswer strin
 	username := remoteUfrag + ":" + localUfrag
 	localPassword := answer.Media[0].GetAttr("ice-pwd")
 	remotePassword := offer.Media[0].GetAttr("ice-pwd")
-	pc.iceAgent = ice.NewAgent(username, localPassword, remotePassword)
+	pc.iceAgent.Configure(username, localPassword, remotePassword)
 
 	return answer.String(), nil
 }
@@ -164,9 +161,6 @@ func (pc *PeerConnection) SetRemoteDescription(sdpOffer string) (sdpAnswer strin
 // Add remote ICE candidate from an SDP candidate string. An empty string denotes the end of
 // remote candidates.
 func (pc *PeerConnection) AddRemoteCandidate(desc string) error {
-	if pc.iceAgent == nil {
-		panic("Must call SetRemoteDescription before AddRemoteCandidate")
-	}
 	return pc.iceAgent.AddRemoteCandidate(desc)
 }
 

